src/lib/zkapp.ts

```diff
@@ -93,84 +110,277 @@ function method<T extends SmartContract>(
     paramTypes,
     SelfProof
   );
+  if (isAsFields(returnType)) methodEntry.returnType = returnType;
   ZkappClass._methods ??= [];
+  let methodIndex = ZkappClass._methods.length;
   ZkappClass._methods.push(methodEntry);
   ZkappClass._maxProofsVerified ??= 0;
   ZkappClass._maxProofsVerified = Math.max(
     ZkappClass._maxProofsVerified,
     methodEntry.proofArgs.length
   );
   let func = descriptor.value;
-  descriptor.value = wrapMethod(func, ZkappClass, methodEntry);
+  descriptor.value = wrapMethod(func, ZkappClass, methodEntry, methodIndex);
 }
 
+let smartContractContext =
+  Context.create<{ this: SmartContract; methodCallDepth: number }>();
+
 // do different things when calling a method, depending on the circumstance
 function wrapMethod(
   method: Function,
   ZkappClass: typeof SmartContract,
-  methodIntf: MethodInterface
+  methodIntf: MethodInterface,
+  methodIndex: number
 ) {
   return function wrappedMethod(this: SmartContract, ...actualArgs: any[]) {
     cleanStatePrecondition(this);
-    if (inCheckedComputation()) {
-      // important to run this with a fresh party everytime, otherwise compile messes up our circuits
-      // because it runs this multiple times
-      let [, result] = Mina.currentTransaction.runWith(
-        {
-          sender: undefined,
-          parties: [],
-          fetchMode: inProver() ? 'cached' : 'test',
-          isFinalRunOutsideCircuit: false,
-        },
+    if (!smartContractContext.has()) {
```

Comment:
This function -- `wrappedMethod` -- is what always gets called when you call any smart contract method. It is where most of the new logic lives. The function is now quite complex!

This first, large if branch is handling the case that was supported already -- calling a zkApp method when you're _not_ inside another zkApp method.